source/openpulse/openpulse/parser.py

```diff
@@ -0,0 +1,264 @@
+    span,
+    QASMNodeVisitor,
+    _visit_identifier,
+    _raise_from_context,
+)
+from openqasm3.antlr.qasm3Parser import qasm3Parser
+
+from .antlr.openpulseLexer import openpulseLexer
+from .antlr.openpulseParser import openpulseParser
+from .antlr.openpulseParserVisitor import openpulseParserVisitor
+from . import ast
+
+
+def parse(input_: str) -> ast.Program:
+    """
+    Parse a complete OpenPulse program from a string.
+
+    :param input_: A string containing a complete OpenQASM 3 program.
+    :return: A complete :obj:`~ast.Program` node.
+    """
+    lexer = openpulseLexer(InputStream(input_))
+    stream = CommonTokenStream(lexer)
+    parser = openpulseParser(stream)
+
+    tree = parser.program()
+
+    return OpenPulseNodeVisitor().visitProgram(tree)
+
+
+class OpenPulseNodeVisitor(openpulseParserVisitor):
+    """Base class for the visitor of the OpenPulse AST."""
+
+    def __init__(self):
+        # A stack of "contexts", each of which is a stack of "scopes".  Contexts
+        # are for the main program, gates and subroutines, while scopes are
+        # loops, if/else and manual scoping constructs.  Each "context" always
+        # contains at least one scope: the base ``ParserRuleContext`` that
+        # opened it.
+        self._contexts: List[List[ParserRuleContext]] = []
+
+    @contextmanager
+    def _push_context(self, ctx: ParserRuleContext):
+        self._contexts.append([ctx])
+        yield
+        self._contexts.pop()
+
+    @contextmanager
+    def _push_scope(self, ctx: ParserRuleContext):
+        self._contexts[-1].append(ctx)
+        yield
+        self._contexts[-1].pop()
+
+    def _current_context(self):
+        return self._contexts[-1]
+
+    def _current_scope(self):
+        return self._contexts[-1][-1]
+
+    def _current_base_scope(self):
+        return self._contexts[-1][0]
+
+    def _in_global_scope(self):
+        return len(self._contexts) == 1 and len(self._contexts[0]) == 1
+
+    def _in_gate(self):
+        return isinstance(self._current_base_scope(), openpulseParser.GateStatementContext)
+
+    def _in_subroutine(self):
+        return isinstance(self._current_base_scope(), openpulseParser.DefStatementContext)
+
+    def _in_loop(self):
+        return any(
+            isinstance(
+                scope, (openpulseParser.ForStatementContext, openpulseParser.WhileStatementContext)
+            )
+            for scope in reversed(self._current_context())
+        )
+
+    def _in_defcal(self):
+        return isinstance(self._current_base_scope(), openpulseParser.DefcalStatementContext)
+
+    @span
+    def _visitPulseType(self, ctx: openpulseParser.ScalarTypeContext):
+        if ctx.WAVEFORM():
+            return ast.WaveformType()
+        if ctx.PORT():
+            return ast.PortType()
+        if ctx.FRAME():
+            return ast.FrameType()
+
+    @span
+    def visitArrayLiteral(self, ctx: openpulseParser.ArrayLiteralContext):
+        array_literal_element = (
```

Comment:
Unfortunately, we had to duplicate a few methods from openqasm3.parser.QASMNodeVisitor because of the instance check. The instances, such as `openpulseParser.ExpressionContext` below and the counterparts in `qasm3Parser` are independently generated from Antlr.

Reply:
It's horribly hacky, but we _can_ get around this without duplicating the code here.  This parser is so coupled to the internal state of the OQ 3 one that if an update would break the hack, it'd break the rest of the parser anyway.

The principle is to monkey-patch the internals of the `openqasm3` ANTLR-generated Python parser to modify the instance checks:

```python
class OpenPulseVersion:
    pass

class FakeMeta(type):
    @classmethod
    def __instancecheck__(cls, instance):
        return isinstance(instance, OpenPulseVersion) or super().__instancecheck__(instance)
    @classmethod
    def __subclasscheck__(cls, other):
        return issubclass(other, OpenPulseVersion) or super().__subclasscheck__(other)

class FakeOpenQASMVersion(openqasm3.antlr.qasmParser.OpenQASMVersion, metaclass=FakeMeta):
    pass

openqasm3.antlr.qasmParser.OpenQASMVersion = FakeOpenQASMVersion
```

Reply:
Just to finish off my thought, because I apparently I pressed submit too soon:

That mutates the actual ANTLR-generated parser (module) to override a context object with a subclass of it, whose entire purpose is to override the behaviour of `isinstance` and `issubclass` so that the new OpenPulse forms type-check as if they were really the OpenQASM types.  Since OpenPulse isn't modifying them from ANTLR and they only appear differently because ANTLR has generated two distinct Python modules, all the functionality is actually identical.

A second alternative is for us to patch the Python output of ANTLR so that the relevant OpenPulse context objects are stubs that directly inherit from the OpenQASM 3 version.  That's perhaps going to be a bit trickier to do, but it's plausible, since we fix to pretty tight versions of the ANTLR runtime.